geode-core/src/distributedTest/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceConstraintsDistributedTest.java

```diff
@@ -0,0 +1,595 @@
+    });
+  }
+
+  /**
+   * The test registers a {@link QueryObserver} that changes the installed
+   * {@link MethodInvocationAuthorizer} to the default one (denies everything) as soon as the query
+   * starts, and makes sure the query succeeds (it is no affected by the authorizer change).
+   * The query is executed from the server.
+   */
+  @Test
+  @Parameters({"LOCAL", "REPLICATE", "PARTITION"})
+  @TestCaseName("[{index}] {method}(RegionType:{0})")
+  public void queriesInFlightExecutedByServersShouldNotBeAffectedWhenMethodAuthorizerIsChanged(
+      RegionShortcut regionShortcut) {
+    String regionName = testName.getMethodName();
+    createAndPopulateRegion(regionName, regionShortcut);
+    String queryString = "<TRACE> SELECT object.getName() FROM /" + regionName + " object";
+
+    server.invoke(() -> {
+      // Set test query observer.
+      assertThat(ClusterStartupRule.getCache()).isNotNull();
+      InternalCache internalCache = ClusterStartupRule.getCache();
+      TestQueryObserver queryObserver =
+          new TestQueryObserver(RestrictedMethodAuthorizer.class.getName(), Collections.emptySet());
+      QueryObserverHolder.setInstance(queryObserver);
+
+      // Execute query.
+      Query query = internalCache.getQueryService().newQuery(queryString);
+      @SuppressWarnings("unchecked")
+      SelectResults<QueryObject> result = (SelectResults<QueryObject>) query.execute();
+      assertThat(queryObserver.invocations.get()).isEqualTo(1);
+      assertThat(result.size()).isEqualTo(ENTRIES);
+
+      // Execute query again, it should fail as the restricted authorizer has been installed.
+      QueryObserverHolder.reset();
+      Query newQuery = internalCache.getQueryService().newQuery(queryString);
+      assertThatThrownBy(newQuery::execute)
+          .isInstanceOf(NotAuthorizedException.class)
+          .hasMessage(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "getName");
+    });
+  }
+
+  /**
+   * The test creates an index with a method invocation as part of the expression, changes
+   * the {@link MethodInvocationAuthorizer} to a custom one that still allows the methods part of
+   * the index expression and executes a region operation from the client that would cause an
+   * index mapping removal/addition.
+   * The operation should succeed, the index should still be valid and no errors should be logged.
+   */
+  @Test
+  @Parameters(method = "getRegionTypeAndOperation")
+  @TestCaseName("[{index}] {method}(RegionType:{0};Operation:{1})")
+  public void indexesShouldNotBeAffectedByMethodAuthorizerChangeAfterRegionOperationOnClientWhenIndexedExpressionContainsMethodsAllowedByTheNewAuthorizer(
+      RegionShortcut regionShortcut, Operation operation) {
+    String regionName = testName.getMethodName();
+    createAndPopulateRegion(regionName, regionShortcut);
+    QueryObject testValue = new QueryObject(999, "name_999");
```

Comment:
Would it be possible to have this `QueryObject` be a class field? The values passed to its constructor seem to be arbitrary so it should be fine for them to be the same in every test case, and it would avoid having to create a new one for every test case.

Reply:
Done!.